pkg/controller/source/source.go

```diff
@@ -46,18 +52,113 @@ type Source interface {
 	Start(eventhandler.EventHandler, workqueue.RateLimitingInterface, ...predicate.Predicate) error
 }
 
+var _ Source = &ChannelSource{}
+
 // ChannelSource is used to provide a source of events originating outside the cluster
 // (eh.g. GitHub Webhook callback).  ChannelSource requires the user to wire the external
 // source (eh.g. http handler) to write GenericEvents to the underlying channel.
-type ChannelSource chan event.GenericEvent
+type ChannelSource struct {
+	// once ensures the event distribution goroutine will be performed only once
+	once sync.Once
+	// source is the source channel to fetch GenericEvents
+	source chan event.GenericEvent
+	// stop is to end ongoing goroutine, and close the channels
+	stop <-chan struct{}
+	// dest is the destination channels of the added event handlers
+	dest []chan event.GenericEvent
+	// destLock is to ensure the destination channels are safely added/removed
+	destLock sync.Mutex
+}
+
+var _ inject.SourceChannel = &ChannelSource{}
+
+// InjectSourceChannel is internal should be called only by the Controller.
+// It is used to inject the source event channel initialized by the ControllerManager.
+func (cs *ChannelSource) InjectSourceChannel(c chan event.GenericEvent) error {
```

Comment:
What do you think of having the user provide the channel when creating the object (like Type on KindSource)?

Private Injected fields are initialized by the ControllerManager.  Public fields should be supplied by the user.

Reply:
I'd I haven't got  enough knowledge on how users will use it.
I think you have more say over this point, so let's make it the way you suggested :)